src/candidate_path.rs

```diff
@@ -1,153 +1,114 @@
 use crate::cfg::ControlFlowGraph;
-use petgraph::graph::NodeIndex;
-use petgraph::visit::EdgeRef;
+use petgraph::{
+    algo::{all_simple_paths, has_path_connecting},
+    graph::NodeIndex,
+    Direction,
+};
 use riscv_decode::Instruction;
 
-pub struct CandidatePath<'a> {
-    pub root: NodeIndex, // instruction we want to evaluate (exit syscall, division with a potential 0 as divisor, ... )
-    pub path: ControlFlowGraph, // actual candidate path from the root to instuctions with no incoming edges and read syscalls
-    pub alternative_roots: Vec<NodeIndex>, // alternative roots for generating new candidate paths from the same control-flow graph (relevant for CandidatePath::next())
-    pub cfg: &'a ControlFlowGraph, // control-flow graph which this candidate path is extracted from
+pub type CandidatePath = (Vec<Instruction>, Vec<bool>);
+
+fn is_exit_node(g: &ControlFlowGraph, idx: NodeIndex) -> bool {
+    g.neighbors_directed(idx, Direction::Outgoing).count() == 0
+        && has_path_connecting(g, NodeIndex::new(0), idx, None)
 }
 
-#[allow(dead_code)]
-impl<'a> CandidatePath<'a> {
-    // may return a reference to a candidate path with a new path and root field
-    pub fn next(&'a mut self) -> Option<&mut CandidatePath> {
-        if let Some(root) = self.alternative_roots.pop() {
-            self.root = root;
-            self.path = ControlFlowGraph::new();
-            self.compute_candidate_path(self.root);
-            Some(self)
-        } else {
-            None
-        }
-    }
+// TODO: Remove all these temporary data structures (collect..)
+pub fn create_candidate_paths<'a>(cfg: &'a ControlFlowGraph) -> Vec<CandidatePath> {
+    cfg.node_indices()
+        .filter(|i| is_exit_node(cfg, i.clone()))
```

Comment:
Wouldn't it make sense not to copy but to refer `i`?

Also clippy warns: using `clone` on a `Copy` type (try dereferencing it: `*i`)